### PR TITLE
Add provision to handle `external` declaration

### LINKED ISSUE #1753: [SciPy][Linalg]: segmentation fault on `external`
```fortran

        subroutine iddr_rsvd(matvect)

        implicit none
        external matvect

        end
```
> LFortran
```console
lfortran --no-warnings --fixed-form --implicit-typing --implicit-interface --show-asr /Users/pgoswami/repos/scipy/scipy/linalg/src/id_dist/src/iddr_rsvd.f
zsh: segmentation fault  lfortran --no-warnings --fixed-form --implicit-typing --implicit-interface
```

### LINKED ISSUE #1776: [SciPy]: This must throw error, it throws seg fault instead
```fortran
subroutine iddr_rsvd(matvect)
   implicit none
end
```

> LFortran

```console
% lfortran ./tests/temp.f90 --show-asr --implicit-interface --implicit-typing
zsh: segmentation fault  lfortran ./tests/temp.f90 --show-asr --implicit-interface --implicit-typing
```

### PR DESCRIPTION
**Pull request by Pranavchiku**
Fixes #1753. Fixes #1776.

### CODE REVIEW

**Comment by Pranavchiku** [comment]
Minimal reproducible error for minpack failure:
```fortran
subroutine sub(sub2)
   external :: sub2
   call sub2(x)
end subroutine
```

**Comment by Pranavchiku** [comment]
This resolves the previous issue, but I guess the test still won't pass because there is some divergence in output of minpack. I will take a detailed look tomorrow morning my time.

**Comment by Pranavchiku** [comment]
Nice, the tests passed!

File: build1.sh [hunk 1]
```diff
@@ -7,7 +7,7 @@ cmake \
     -DCMAKE_BUILD_TYPE=Debug \
     -DWITH_LLVM=yes \
     -DLFORTRAN_BUILD_ALL=yes \
-    -DWITH_STACKTRACE=yes \
+    -DWITH_STACKTRACE=no \
```
Comment (certik) [inline build1.sh:10]: Change this back.

File: src/lfortran/semantics/ast_common_visitor.h [hunk 2]
```diff
@@ -1532,9 +1534,54 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
+                                    assgnd_access[sym] = ASR::accessType::Public;
+                                    SymbolTable *parent_scope = current_scope;
+                                    current_scope = al.make_new<SymbolTable>(parent_scope);
+                                    ASR::ttype_t *type;
+                                    if (sym_) {
+                                        type = ASRUtils::symbol_type(sym_);
+                                    } else if (compiler_options.implicit_typing) {
+                                        type = implicit_dictionary[sym];
+                                        if (!type) {
+                                            // currently hardcoing type to real*4
```
Comment (certik) [inline src/lfortran/semantics/ast_common_visitor.h:1548]:
```suggestion
                                            // currently hardcoding type to real*4
```

File: src/lfortran/semantics/ast_common_visitor.h [hunk 3]
```diff
@@ -1532,9 +1534,54 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
+                                            type = ASRUtils::TYPE(ASR::make_Real_t(al, x.base.base.loc, 4));
+                                        }
+                                    } else {
+                                        // currently hardcoing type to real*4
```
Comment (certik) [inline src/lfortran/semantics/ast_common_visitor.h:1552]:
```suggestion
                                        // currently hardcoding type to real*4
```

File: src/lfortran/semantics/ast_common_visitor.h [hunk 4]
```diff
@@ -1532,9 +1534,54 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
+                                        type = ASRUtils::TYPE(ASR::make_Real_t(al, x.base.base.loc, 4));
```
Comment (certik) [inline src/lfortran/semantics/ast_common_visitor.h:1553]: Put a TODO comment here explaining what needs to be done -- should the type be looked up based on the implicit typing thing?

**Comment by certik** [comment]
I left some comments.

File: src/lfortran/semantics/ast_common_visitor.h (hunk 4 above)
Comment (Pranavchiku) [reply src/lfortran/semantics/ast_common_visitor.h:1553]: This is the case where we have no information about type, this is what gfortran does:
```console
% gfortran ./tests/external1.f90 -fdump-parse-tree
Namespace: A-Z: (UNKNOWN 0)
procedure name = iddr_rsvd
  symtree: 'iddr_rsvd'   || symbol: 'iddr_rsvd'    
    type spec : (UNKNOWN 0)
    attributes: (PROCEDURE  SUBROUTINE)
    Formal arglist: matvect
  symtree: 'matvect'     || symbol: 'matvect'      
    type spec : (UNKNOWN 0)
    attributes: (PROCEDURE  EXTERNAL DUMMY SUBROUTINE ARRAY-OUTER-DEPENDENCY)

  code:
  
------------------------------------------
```

Comment (Pranavchiku) [reply src/lfortran/semantics/ast_common_visitor.h:1553]: I will specify this instead writing `currently hardcoing type to real*4`

**Comment by certik** [comment]
Thanks.

**Pull request merged**